subprojects/testing-base/src/integTest/groovy/org/gradle/testing/TestListenerBuildOperationAdapterIntegrationTest.groovy

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.testing
+
+import org.gradle.api.tasks.testing.ExecuteTestBuildOperationType
+import org.gradle.api.tasks.testing.TestOutputBuildOperationType
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.integtests.fixtures.BuildOperationsFixture
+import org.gradle.integtests.fixtures.TestResources
+import org.gradle.internal.operations.trace.BuildOperationRecord
+import org.junit.Rule
+
+class TestListenerBuildOperationAdapterIntegrationTest extends AbstractIntegrationSpec {
```

Comment:
Ideally, we would be reusing our existing strategies for asserting on our understanding of the test execution, which is based around `TestExecutionResult`. We already have a bunch of tests that assert how we represent different quirky test arrangements in our model, of which these new operations are now effectively part. 

In practice though, this may not be worth it. It would be some work to have TestExecutionResult transparently assert on the event stream, if it is at all possible without drastic changes.

My overarching concern here is that functional tests for this new feature are patchy. We either need to make that not the case, our base our QA on implicit comprehensive coverage (e.g. we assume that correct HTML test result reports translates into correct emission of operation notifications).

Reply:
one alternative would be to generate the trace file and use that to gather the results in a TextExecutionResult. on the plus side it would be way easier to get TestExecutionResult transparently with that. The downside is that it comes with quite a cost.